src/tun.rs

```diff
@@ -120,6 +120,52 @@ impl Tun {
         Ok(iface)
     }
 
+    /// Receives a packet from the Tun/Tap interface
+    ///
+    /// This method takes &self, so it is possible to call this method concurrently with other methods on this struct.
+    pub async fn recv(&self, buf: &mut [u8]) -> io::Result<usize> {
+        loop {
```

Comment:
using a loop inside an async method is not a good idea, please change it to:

```rust
pub fn poll_recv(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
        buf: &mut ReadBuf<'_>,
    ) -> task::Poll<io::Result<()>> {
    // ...
}
```

Reply:
Hello @yaa110 ,

Could you explain a bit on what you mean by "using a loop inside an async method is not a good idea"?

The new APIs are intended for concurrent calling by multiple tasks at the same time. If I implement `poll_recv` as what was implemented above, then I would have to use https://docs.rs/tokio/1.12.0/tokio/io/unix/struct.AsyncFd.html#method.poll_read_ready which specifically mentions:

> Note that on multiple calls to poll_read_ready or poll_read_ready_mut, only the Waker from the Context passed to the most recent call is scheduled to receive a wakeup. (However, poll_write_ready retains a second, independent waker).
> ...
> This method is intended for cases where creating and pinning a future via readable is not feasible. Where possible, using readable is preferred, as this supports polling from multiple tasks at once.

Also, I followed the API of the `UDPSocket.try_recv()`, which uses `&mut [u8]` instead of `ReadBuf` for output buffer and avoids having to do extra heap allocations for each read.

Reply:
your implementation blocks the execution of program. An async method should not be blocking.

Reply:
Hello @yaa110,

This method, when called, will return a future that can be either `await`ed by the caller, or be polled with things like `tokio:spawn` or even spin off to run somewhere else using `tokio::spawn`. It does not imply the caller will be blocked, the caller is free to determine whether to block on the returned future or not. This is the same API interface as https://docs.rs/tokio/1.12.0/tokio/net/struct.UdpSocket.html#method.recv where the caller can choose to block on `recv` or not. Note that the implementation of `UDPSocket::recv` from Tokio also used `.await` in the function body.

Reply:
I know what the return type is. In your implementation the method is blocked after awaiting. However, it should not block when it is not ready.

Reply:
Please see: https://docs.rs/tokio/1.11.0/src/tokio/net/udp.rs.html#670-675
and https://github.com/tokio-rs/tokio/blob/master/tokio/src/io/driver/registration.rs#L249-L260

Note that the `recv` method of `UDPSocket` was implemented the same way, i.e. `await` in a loop.